for_fedora/man/shinken.8
.TH shinken 8 "July 7, 2014" "Shinken User Manuals"
.SH NAME
.PP
shinken - Shinken packs and modules manager
.SH SYNOPSIS
.PP
shinken [OPTION]... \f[I]FILE\f[] 
.SH DESCRIPTION
.PP
Shinken packs and modules manager
.PP
Shinken is a usefull tool to manage your Shinken packs and modules and 
can be used to download packs and modules from \f[I]http://shinken.io\f[] and 
also to publish yours.
.SH OPTIONS
.TP
.B --version
Show version number and exit
.RS
.RE
.TP
.B -A \f[I]API_KEY\f[], --api-keys \f[I]API_KEY\f[]
Your API key for uploading the package to the
Shinken.io website. If you don't have one, please go
to your account page
.RS
.RE
.TP
.B -l, --list
List available commands
.RS
.RE
.TP
.B --init
Initialize/refill your shinken.ini file (default home directory)
.RS
.RE
.TP
.B -D
Enable the debug mode
.RS
.RE
.TP
.B -c \f[I]INICONFIG\f[], --config \f[I]INICONFIG\f[]
Path to your shinken.ini file.
.RS
.RE
.TP
.B -h, --help
Print detailed help screen.
.RS
.RE
.SH AUTHORS
David Hannequin.
Michael Leinartas.


Changelog
########################
Shinken ChangeLog
########################

2.4   - 04/05/2015
-------------------
	CORE ENHANCEMENT
	Add: Hosts - Service_includes feature. Similar to service_excludes
	Add: Deprecation in the doc about the discovery part. Should be move to contrib or as a module in next versions
	Add: (by Ddurieux) better logging of failed external modules


	Enh: Actions - Catch stderr output
	Enh: Arbiter -  Unpickle new conf earlier. Prevent a memory issue
	Enh: Config - Remove sudo from restart/reload commands
	Enh: Daemon - Factorization of  get_objects_from_from_queues
	Enh: DaemonsLinks : Move files to object directory
	Enh: Datamanager - Function get_contactgroup
	Enh: Doc - A lot of typo and improvements!
	Enh: Init Scripts - use readlink to avoid failure on symlinks
	Enh: Modulesmanager - Reworking module loading, now try to import in 3 different ways
	Enh: Pep8 : Shinken is now pep8 compliant (enforced by Travis) excepted for 3 specific rules
	Enh: Scheduler - Use iterator instead of list when necessary. Reduce memory consumption
	Enh: Setup - Refactoring to be virtualenv compatible
	Enh: Tests - Stabilization


	Del: Check Shinken script. It has its own repository
	Del: Windows files from 1.4 version


	CORE FIXES
	Fix: Config Parsing - Arbiter module auto creation
	Fix: Config Parsing - Infinite recursion loop in template
	Fix: Config Parsing - Inheritance of custom variable
	Fix: Config Parsing - Raise error properly in ArbiterLink object
	Fix: Config Parsing - Type properties in Contact, NotificationWay and Arbiter objects
	Fix: Escalations - escalated parameter in notification object and macro $NOTIFICATIONISESCALATED$ are set correctly
	Fix: Event Handlers - Global event handlers parsing and launching
	Fix: External Command - Internal argument for CHANGE_RETRY_HOST_CHECK_INTERVAL command
	Fix: Realm - Bailout when there are two default realm
	Fix: (reported by:Stéphane Loeuillet) init.d - Unset http(s)_proxy env variables in the init.d script (curl was using them by default)
	Fix: (reported by: Arthur Lutz) CLI - manage shinken install call without arguments
	Fix: (reported by mohierf) STATS - fix statsd format.
	Fix: (reported by: Stéphane Loeuillet) Config Parsing - manage , case at the end of members list (hostgroups & contactgroups)
	Fix: (reported by: Stéphane Loeuillet) Config Parsing - host check_period was mandatory with some parameters, now is missing always true period.
	Fix: + char in servicegroups




2.2   - 16/01/2015
-------------------
	CORE ENHANCEMENT
	Add : Bottle - HTTP_X_FORWARDED_PROTO support (for WebUI)
	Add : BpRules - Expands bp_rules t flag from tags
	Add : BpRules - Use standard macros expansion in bp_rules output
	Add : Broks - new type unknown_check_result_brok (included in the previous add)
	Add : CLI - update command for the shinken.io install
	Add : Contact - new parameter expert. For UI purpose
	Add : Escalations - Time properties for host and service escalations
	Add : External Command - Implement SCHEDULE_HOSTGROUP_HOST_DOWNTIME and SCHEDULE_HOSTGROUP_SVC_DOWNTIME
	Add : External Commands - Reload-config function
	Add : External Commands - support for reset modify attribute
	Add : HTTP API - get_start_time call returning timestamp
	Add : HTTP API - Inner stats can be sent to shinken.io
	Add : LogEvent - Output key to properties dict
	Add : LogEvent - Pattern for FLAPPING events
	Add : Receiver - Accept passive results from host or service not in configuration
	Add : Realms - Handle properly multi level of realms
	Add : SchedulingItem - callbacks for HOSTDOWNTIME and SERVICEDOWNTIME macros
	Add : SchedulingItem - Parameter to remove host-service dependency
	Add : Service Escalation - wildcard service_description
	Add : Service - method  get_service_tags (for WebUI)
	Add : Snapshots  - Feature to get more output when we have a bad state
	
	Del : Move WebUI image and template to the good repo
	Del : Pack Distribution tuning (hosts where fixd on the same scehduler between reload) - causing more damage in load balancing that it solve problems

	Enh : CLI now exit with a real system return code when fail to install
	Enh : Config - Module name formatting in config files
	Enh : Config parsing - Refactoring to pythonize at object creation. Attribute have correct type very early.
	Enh : Config parsing - Refactoring to reduce parse time, and scope code to ease future evolutions
	Enh : Daemon - Refactor bailout and edit message
	Enh : Datarange - Better datarange parsing
	Enh : Deamon - additional call to  os.setgroups to drop privileges
	Enh : Doc - A lot of typo and improvments!
	Enh : External Command - Moved manager initialization to init
	Enh : Imports - use importlib when available to import module properly.
	Enh : Init scripts - Refactoring script. Add LSB standard functions
	Enh : Logger - Only use time rotated file handler for regular files
	Enh : Logger - Refactoring logger class to inherit from standard logger class.
	Enh : Logger - Use lazy logger syntax to enhance performances
	Enh : Perfdata - Better internal parsing
	Enh : Realm - Code refactoring
	Enh : Scheduler - use a lock to acces put_result
	Enh : SchedulingItem - use return_code attribute to remember plugin return code so that it is available in broks.
	Enh : Tests - better Jenkins and Travis integration and Coverall support
	Enh : Tests - Code cleaning
	Enh : Tests - Use assert function from unittest2
	Enh : Triggers - Try except statement when executing them


	CORE FIXES
	Fix : BpRules - '+' was causing segfault in some cases
	Fix : Broker - Crash when sub process goes down
	Fix : Config parsing - Unknown members list is common to all items
	Fix : Daemon - Proper https URL
	Fix : Daemons - Timeout and data timeout for scheduler connexion
	Fix : Daemon - When replacing a daemon kill all processes in the group instead of one pid
	Fix : Datamanager -  Service totals and host totals computation
	Fix : Discovery - Linux hosts with recent nmap versions

contrib/libexec/eue/features/support/shinken.rb
require "rubygems"
require 'cucumber/formatter/console'
require 'date'
require 'mongo'
formater_path = File.expand_path(File.dirname(__FILE__)+"/../../features/support")        
require "#{formater_path}/commons.rb"


module Eue
    class Shinken < Commons

      def initialize(step_mother, io, options={})

          base_path = File.expand_path(File.dirname(__FILE__)+"/../../")        
          project = ""

          ARGV.each do | element |
              if element.index(/\.feature/)
                  fullname=element.split("/")[element.split("/").length-1]
                  project = fullname.split(".")[0]
                  break
              end
          end

          if project == ""
              exit 2
          end

          @@params = IniFile.new(base_path+"/#{project}.ini", :parameter => '=')
          @@connection = nil
          @@db = nil
          @@collection = nil
          @@video = ""
          connect()
      end

      def after_feature(feature)
          if @@params["media"]["capturevideo"].to_i == 1
            media = @@params["media"]["path"]
            @@video = "#{@@start_time.to_i}_#{normalizestring(normalizestring(@@params["application"]["name"]))}_#{normalizestring(@@feature)}.#{@@params['media']['videoextension']}"
            begin
              $headless.video.stop_and_save("#{media}/#{@@video}")
            rescue

            end
            videodata = File.open("#{media}/#{@@video}")
            dbgrid = @@connection.db(@@params["mongo"]["dbgrid"])        
            fsgrid = Mongo::GridFileSystem.new(dbgrid)
            fsgrid.open(@@video, "w") do |f|
              f.write videodata
            end         
            #File.delete("#{media}/#{@@video}")
          end
          print_summary()
      end

      private

      def connect()
        begin
          @@connection = Mongo::Connection.new(@@params["mongo"]["host"],@@params["mongo"]["port"].to_i)
          # puts "Connected to %s %s"  % [@@params["mongo"]["host"],@@params["mongo"]["port"]]
        rescue
          puts "Unable to connect to mongodb server %s %s" % [@@params["mongo"]["host"],@@params["mongo"]["port"]]
          exit 2
        end

        begin
          @@db = @@connection.db(@@params["mongo"]["db"])
          # puts "%s db opened" % @@params["mongo"]["db"]
        rescue
          puts "Unable to connect to database %s" % @@params["mongo"]["db"]
          exit 2
        end

        begin
          @@collection = @@db.collection(@@params["mongo"]["collection"])
          # puts "%s collection opened" % @@params["mongo"]["collection"]
        rescue
          puts "Unable to open collection %s" % @@params["mongo"]["collection"]
          exit 2
        end

      end

      def insert_document(document)
        id = @@collection.insert(document)
        return id
      end

      def print_summary

        key = "%s.%s.%s.%s.%s.%s" % [ 
                  @@start_time.to_i.to_s,
                  normalizestring(@@params["application"]["name"]),
                  normalizestring(@@feature),
                  normalizestring(@@params["robot"]["os"]),
                  normalizestring(@@params["browser"]["name"]),
                  normalizestring(@@params["robot"]["localization"])
        ]

        document = {
          "key" => key,
          "application" => @@params["application"]["name"],
          "application_code" => normalizestring(@@params["application"]["name"]),
          "feature" => @@feature,
          "start_time" => @@params["mongo"]["dateformat"] == "native" ? @@start_time : @@start_time.to_i,
          "description" => @@description,
          "platform" => {
            "os" => @@params["robot"]["os"],
            "localization" => @@params["robot"]["localization"],
            "browser" => @@params["browser"]["name"]
          },
          "dbmedia" => @@params["mongo"]["dbmedia"],
          "screenshots_path" => @@params["media"]["path"],
          "video" => @@video,
          "scenarios" => @@scenarios,
          "robot" => @@params["robot"]["name"]
        }

        id = insert_document(document)

        ##puts @@document.to_json.to_s.gsub("{","{\n").gsub("}","\n}\n").gsub(",",",\n")
        output_shinken(key)
      end

      def take_screenshot(feat, scenario, step, scenario_index, step_index)
        if step.to_s != ""
          feature,*descritption = feat.split("\n")
          screenshot_file = "#{@@cur_scenario_start.to_i}_#{normalizestring(feature)}_%04d_#{normalizestring(scenario)}_%04d_#{normalizestring(step)}.png" % [scenario_index,step_index]
        else
          feature,*descritption = feat.split("\n")
          screenshot_file = "#{@@cur_scenario_start.to_i}_#{normalizestring(feature)}_%04d_#{normalizestring(scenario)}.png" % scenario_index
        end          
        screenshot = "#{@@params["media"]["path"]}/#{screenshot_file}"
        Browser.driver.save_screenshot(screenshot)

        image = File.open(screenshot)
        dbgrid = @@connection.db(@@params["mongo"]["dbgrid"])        
        fsgrid = Mongo::GridFileSystem.new(dbgrid)
        fsgrid.open(screenshot_file, "w") do |f|
          f.write image
        end         
        File.delete(screenshot)
        return screenshot_file      
      end

    end
end

